src/neovim.rs

```diff
@@ -44,6 +45,69 @@ type ResponseResult = Result<Result<Value, Value>, Arc<DecodeError>>;
 
 type Queue = Arc<Mutex<Vec<(u64, oneshot::Sender<ResponseResult>)>>>;
 
+/// The current state of redraw notifications. Redraws must be dispatched
+/// sequentially in the order that they were received, as each redraw
+/// notification assumes it starts with the result from the previous redraw. See
+/// `:help ui-events` in Neovim for more info.
+struct RedrawQueue {
+  /// Whether there's a future already active that's flushing redraw
+  /// notifications
+  is_active: bool,
+  /// The queue of pending redraw notifications
+  queue: VecDeque<Vec<Value>>,
+}
+
+impl RedrawQueue {
+  fn new() -> Self {
+    RedrawQueue {
+      is_active: false,
+      queue: VecDeque::<Vec<Value>>::new(),
+    }
+  }
+}
+
+async fn queue_redraw<H, W>(
+  queue: &Arc<Mutex<RedrawQueue>>,
+  handler: H,
+  nvim: Neovim<H::Writer>,
+  params: Vec<Value>,
+)
+where
+  W: AsyncWrite + Send + Unpin + 'static,
+  H: Handler<Writer = W>
+{
+  let mut guard = queue.lock().await;
+
+  guard.queue.push_front(params);
+  if guard.is_active {
+    return;
+  }
+  guard.is_active = true;
+  drop(guard);
+
+  let queue = queue.clone();
+  let handler_c = handler.clone();
+  let neovim = nvim.clone();
+  handler.spawn(async move {
```

Comment:
This probably needs a `+ create::Spawner` bound, so that you can actually use `handler::spawn`.